robotframework-ls/src/robotframework_ls/impl/semantic_tokens.py

```diff
@@ -97,9 +108,101 @@ def extract_library_token_from_keyword(keyword_token):
+        )
+        bdd_statement_token.index = RF_TOKEN_TYPE_TO_TOKEN_TYPE_INDEX[KEYWORD]
+    return bdd_statement_token
+
+
+def extract_embedded_variable(keyword_token):
+    import re
+
+    embedded_variable_token = None
+    result = re.match(r"([$&@]{[^}]*?})", keyword_token.value)
+    if result:
+        embedded_variable_value = result.group(1)
+        embedded_variable_length = len(embedded_variable_value)
+        trailing_whitespace = 0
+        if embedded_variable_length < len(keyword_token.value):
+            if keyword_token.value[embedded_variable_length].isspace():
+                trailing_whitespace = 1
+        embedded_variable_token = keyword_token.extract_token(
+            "variable", embedded_variable_length, trailing_whitespace
+        )
+        embedded_variable_token.index = TOKEN_TYPE_TO_INDEX[
+            embedded_variable_token.type
+        ]
+    return embedded_variable_token
+
+
+def split_variable_into_subtokens(variable_token):
+    opening_brackets_token = variable_token.extract_token("variableOperator", 2, 0)
+    opening_brackets_token.index = VARIABLE_OPERATOR_INDEX
+    embedded_variable_length = (
+        variable_token.end_col_offset - variable_token.col_offset - 1
+    )
+    embedded_variable_token = variable_token.extract_token(
+        "variable", embedded_variable_length, 0
+    )
+    embedded_variable_token.index = TOKEN_TYPE_TO_INDEX[embedded_variable_token.type]
+    closing_brackets_token = variable_token.extract_token("variableOperator", 1, 0)
+    closing_brackets_token.index = VARIABLE_OPERATOR_INDEX
+    return [opening_brackets_token, embedded_variable_token, closing_brackets_token]
+
+
+def extract_tokens_from_bdd_statement(token_being_parsed):
+    token_has_been_fully_parsed = False
+    list_with_subtokens = []
+    if keyword_contains_embedded_variable(token_being_parsed):
+        token_bdd_statement = extract_part_of_bdd_statement(token_being_parsed)
+        if token_bdd_statement:
+            if token_bdd_statement == token_being_parsed:
+                token_has_been_fully_parsed = True
+            list_with_subtokens.append(token_bdd_statement)
+        token_embedded_variable = extract_embedded_variable(token_being_parsed)
+        if token_embedded_variable:
+            if token_embedded_variable == token_being_parsed:
+                token_has_been_fully_parsed = True
+            list_with_subtokens = list_with_subtokens + split_variable_into_subtokens(
+                token_embedded_variable
+            )
+        if not token_has_been_fully_parsed:
+            tokens_parsed_recursively = extract_tokens_from_bdd_statement(
+                token_being_parsed
+            )
+            list_with_subtokens = list_with_subtokens + tokens_parsed_recursively
+    else:
+        list_with_subtokens = [token_being_parsed]
+    return list_with_subtokens
+
+
+def keyword_contains_embedded_variable(keyword_token):
```

Comment:
This method should be replaced by: `robotframework_ls.impl.text_utilities.contains_variable_text` -- although maybe ideally we should just do the call to `ast_utils.tokenize_variables(token)` just checking if `{` is present (because the tokenization is slow and if there are no tokens it'll just return the initial token and there'll be no variables in it).